libvast/src/arrow_table_slice.cpp

```diff
@@ -1141,4 +1146,171 @@ template class arrow_table_slice<fbs::table_slice::arrow::v0>;
+    return *result;
+  die("failed to finish Arrow subnet column builder");
+}
+
+// TODO: this is copied from
+// experimental_table_slice_builder/enum_column_builder, can this live somewhere
+// shared?
+std::shared_ptr<arrow::Array> make_dict_array(const enumeration_type& et) {
+  arrow::StringBuilder dict_builder{};
+  for (const auto& f : et.fields())
+    if (!dict_builder.Append(std::string{f.name}).ok())
+      die("failed to build Arrow enum field array");
+  if (auto array = dict_builder.Finish(); array.ok()) {
+    return *array;
+  }
+  die("failed to finish Arrow enum field array");
+}
+
+/// Converts an array representing a single column from a previous arrow
+/// format to `arrow::experimental`.
+std::shared_ptr<arrow::Array>
+convert_column(const std::shared_ptr<arrow::Array>& arr, const type& t) {
+  auto f = detail::overload{
+    [&](const auto& t) -> std::shared_ptr<arrow::Array> {
+      VAST_ASSERT(arr->type()->Equals(t.to_arrow_type()));
+      return arr;
+    },
+    [&](const record_type& rt) -> std::shared_ptr<arrow::Array> {
+      // this case handles VAST type `list<record>`
+      const auto& sa = static_pointer_cast<arrow::StructArray>(arr);
+      arrow::ArrayVector children{};
+      std::vector<std::string> field_names{};
+      children.reserve(rt.num_fields());
+      field_names.reserve(rt.num_fields());
+      int index = 0;
+      for (const auto& f : rt.fields()) {
+        const auto& arr = sa->field(index++);
+        children.push_back(convert_column(arr, f.type));
+        field_names.emplace_back(f.name);
+      }
+      auto res = arrow::StructArray::Make(children, field_names);
+      if (!res.ok())
+        die("unable to construct nested struct array");
+      return *res;
+    },
+    [&](const pattern_type&) -> std::shared_ptr<arrow::Array> {
+      return std::make_shared<pattern_type::array_type>(
+        pattern_type::to_arrow_type(), arr);
+    },
+    [&](const address_type&) -> std::shared_ptr<arrow::Array> {
+      return std::make_shared<address_type::array_type>(
+        address_type::to_arrow_type(), arr);
+    },
+    [&](const subnet_type&) -> std::shared_ptr<arrow::Array> {
+      auto ba = static_pointer_cast<arrow::FixedSizeBinaryArray>(arr);
+      VAST_ASSERT(ba->byte_width() == 17);
+      return convert_subnet_array(*ba);
+    },
+    [&](const enumeration_type& et) -> std::shared_ptr<arrow::Array> {
+      auto dict = make_dict_array(et);
+      auto indices = arrow::compute::Cast(*arr, arrow::uint8());
+      auto dict_array = std::make_shared<arrow::DictionaryArray>(
+        et.to_arrow_type()->storage_type(), indices.ValueUnsafe(), dict);
+      return std::make_shared<enumeration_type::array_type>(et.to_arrow_type(),
+                                                            dict_array);
+    },
+    [&](const list_type& lt) -> std::shared_ptr<arrow::Array> {
+      auto la = static_pointer_cast<arrow::ListArray>(arr);
```

Comment:
You never need this as a shared pointer, so why pointer cast over casting to a const-lvalue-ref of the underlying element type?